[dynamo] allow DeviceMesh variable desugar ProcessGroup

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111461

As titled, when calling get_dim_groups, DeviceMeshVariable should be
de-sugared to ProcessGroup Variable in dynamo

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

Differential Revision: [D50400451](https://our.internmc.facebook.com/intern/diff/D50400451)

wanchaol committed:
[dynamo] allow DeviceMesh variable desugar ProcessGroup

As titled, when calling get_dim_groups, DeviceMeshVariable should be
de-sugared to ProcessGroup Variable in dynamo

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50400451).

voznesenskym commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -148,9 +148,36 @@ def is_device_mesh(value):
+        if name == "get_dim_groups":
+            args = [x.as_python_constant() for x in args]
+            kwargs = {k: v.as_python_constant() for k, v in kwargs.items()}
+            dim_groups = self.value.get_dim_groups(*args, **kwargs)
```
This is invoking code, that might potentially invoke user code if someone subclasses DeviceMesh, right?

wanchaol replied:
it should be fine as when we create DeviceMeshVariable we explicitly checked `istype(val, DeviceMesh)`, so it should be safe. I'll add a assertion on type here to further ensure it

voznesenskym commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -148,9 +148,36 @@ def is_device_mesh(value):
+            kwargs = {k: v.as_python_constant() for k, v in kwargs.items()}
+            dim_groups = self.value.get_dim_groups(*args, **kwargs)
+            # desugar the results to ProcessGroupVariables
+            options = VariableTracker.propagate(self)
```
where do you propagate guards from the args and kwargs?

wanchaol replied:
Good point, I'll update

wanchaol committed:
Update on "[dynamo] allow DeviceMesh variable desugar ProcessGroup"


As titled, when calling get_dim_groups, DeviceMeshVariable should be
de-sugared to ProcessGroup Variable in dynamo

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50400451](https://our.internmc.facebook.com/intern/diff/D50400451)

[ghstack-poisoned]

wanchaol committed:
Update on "[dynamo] allow DeviceMesh variable desugar ProcessGroup"


As titled, when calling get_dim_groups, DeviceMeshVariable should be
de-sugared to ProcessGroup Variable in dynamo

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50400451](https://our.internmc.facebook.com/intern/diff/D50400451)

[ghstack-poisoned]
